### PR TITLE
feat(send-notification): map errors to error response

### PR DESCRIPTION
**Pull request by haonliu**
Map errors to error response

### CODE REVIEW

File: packages/service-library/src/web-api/scan-notification-error-codes.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    code: ScanNotificationErrorCodeName;
+    codeId: number;
+    message: string;
+}
+
+export class ScanNotificationErrorCodes {
+    public static internalError: ScanNotificationErrorCode = {
+        code: 'InternalError',
+        codeId: 9001,
+        message: 'The scan engine encountered an internal error.',
```
Comment (devaradhanm) [inline packages/service-library/src/web-api/scan-notification-error-codes.ts]: notification sender

File: packages/service-library/src/web-api/scan-notification-error-codes.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    };
+
+    public static HttpErrorCode: ScanNotificationErrorCode = {
+        code: 'HttpErrorCode',
+        codeId: 9012,
+        message: 'Failed to send scan notification.',
```
Comment (lamaks) [inline packages/service-library/src/web-api/scan-notification-error-codes.ts]: this is too generic error message, not clear how this correlates with Http... should code be just a `NotificationError` instead?

File: packages/service-library/src/web-api/scan-notification-error-codes.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    // This type is part of the REST API client response.
+    // Ensure compatibility when changing this type.
+    code: ScanNotificationErrorCodeName;
+    codeId: number;
+    message: string;
+}
+
+export class ScanNotificationErrorCodes {
+    public static NotificationError: ScanNotificationErrorCode = {
+        code: 'NotificationError',
```
Comment (devaradhanm) [inline packages/service-library/src/web-api/scan-notification-error-codes.ts]: you should also be sending the last notification url response code. (similar to how we do it for scan)

File: packages/web-api/src/converters/scan-response-converter.ts [hunk 4]
```diff
@@ -6,17 +6,14 @@ import { ScanReport, ScanResultResponse } from 'service-library';
 import { OnDemandPageScanResult, OnDemandPageScanRunState, ScanCompletedNotification } from 'storage-documents';
 import { isNullOrUndefined } from 'util';
```
Comment (devaradhanm) [inline packages/web-api/src/converters/scan-response-converter.ts]: isNullOrUndefined is deprecated. use isNil from lodash

File: packages/web-api/src/converters/scan-response-converter.ts [hunk 5]
```diff
@@ -59,7 +56,7 @@ export class ScanResponseConverter {
                         pageResponseCode: pageScanResultDocument.run.pageResponseCode,
                         pageTitle: pageScanResultDocument.run.pageTitle,
                     },
-                    ...notificationResponse,
+                    ...this.getRunCompleteNotificationResponse(pageScanResultDocument.notification),
                 };
                 if (pageScanResultDocument.scannedUrl !== undefined) {
                     scanResultResponse.scannedUrl = pageScanResultDocument.scannedUrl;
```
Comment (devaradhanm) [inline packages/web-api/src/converters/scan-response-converter.ts]: looks like this line doesnt have code coverage

File: packages/web-api/src/converters/scan-response-converter.spec.ts [hunk 6]
```diff
@@ -129,12 +152,11 @@ describe(ScanResponseConverter, () => {
...
         validateConverterShortResult('running', 'running', notificationEnabled);
         validateConverterShortResult('failed', 'failed', notificationEnabled);
-        scanRunErrorConverterMock.verifyAll();
     });
 
-    it('return scan run full form of client result', () => {
-        const pageScanDbResult = getPageScanResult('completed');
-        const responseExpected = getScanResultClientResponseFull('completed');
+    test.each([true, false])('return scan run full form of client result', notificationEnabled => {
+        const pageScanDbResult = getPageScanResult('completed', notificationEnabled);
```
Comment (devaradhanm) [inline packages/web-api/src/converters/scan-response-converter.spec.ts]: add the value of testcase to the test case name to avoid duplicate test name

File: packages/web-api-send-notification-runner/src/sender/notification-sender.ts [hunk 7]
```diff
@@ -63,27 +64,34 @@ export class NotificationSender {
                     this.logger.logInfo(`Notification sent Successfully!, try #${numberOfTries}`);
                     notificationState = 'sent';
                     error = null;
+                    statusCode = 200;
```
Comment (devaradhanm) [inline packages/web-api-send-notification-runner/src/sender/notification-sender.ts]: set statusCode above if else.

File: packages/service-library/src/web-api/api-contracts/scan-result-response.ts [hunk 8]
```diff
@@ -8,7 +9,7 @@ export declare type ReportFormat = 'sarif' | 'html';
 export declare type ScanState = 'pending' | 'pass' | 'fail';
 export declare type RunState = 'pending' | 'accepted' | 'queued' | 'running' | 'completed' | 'failed';
 export declare type NotificationState = 'pending' | 'queued' | 'queueFailed' | 'sent' | 'sendFailed';
-export type NotificationErrorTypes = 'HttpErrorCode';
+export type NotificationErrorTypes = 'NotificationError' | 'InternalError' | 'HttpErrorCode';
```
Comment (devaradhanm) [inline packages/service-library/src/web-api/api-contracts/scan-result-response.ts]: remove NotificationError

File: packages/service-library/src/web-api/scan-notification-error-codes.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// tslint:disable: no-unnecessary-class variable-name
+
+export declare type ScanNotificationErrorCodeName = 'NotificationError' | 'InternalError' | 'HttpErrorCode';
```
Comment (devaradhanm) [inline packages/service-library/src/web-api/scan-notification-error-codes.ts]: remove NotificationError

File: packages/service-library/src/web-api/scan-notification-error-codes.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+export interface ScanNotificationErrorCode {
+    // This type is part of the REST API client response.
+    // Ensure compatibility when changing this type.
+    code: ScanNotificationErrorCodeName;
+    codeId: number;
+    message: string;
+}
+
+export class ScanNotificationErrorCodes {
+    public static NotificationError: ScanNotificationErrorCode = {
```
Comment (devaradhanm) [inline packages/service-library/src/web-api/scan-notification-error-codes.ts]: remove this
